### PR TITLE
Batch update columns using id instead of column_index

### LINKED ISSUE #1167: Batch updating column using `/tables/` api should reference column using `id` instead of `column_index`
## Problem
Batch updating Columns using the `columns` field in the `patch` method of the `/tables/{id}/` does not use `id` of the column rather the column to update is identified based on the position of the column objects. This is not ideal due to the following reasons:
- Similar to the `column_index` which is [deprecated in favour](https://github.com/centerofci/mathesar/issues/839) of `id`.
- Column position could have changed resulting in updating a wrong column

## Proposed solution
The following changes have to be made:

-`PATCH` method of the `/tables/{id}` api which has the following structure 

```
{
    "columns": [
        {
            "name": "id",
            "type": "INTEGER"
        },
        {
            "name": "Center",
            "type": "TEXT"
        }
    ]
}
```

should be replaced with 

```
{
    "columns": [
        {
            "id": 1, // Use column id to reference a column, rather than assuming the column based on its index.
            "name": "id",
            "type": "INTEGER"
        },
        {
            "id": 2,
            "name": "Center",
            "type": "TEXT"
        }
    ]
}
```

- The column id's have to be converted into `attnums` which can be fetched from the column object before sending the column data from the service layer to the database layer(before calling any function from the `db` module)

### PR DESCRIPTION
**Pull request by Anish9901**
Fixes #1167 

Currently, batch updating columns was done based on the position of the column object which could be unreliable, to fix this, we make use of the `id` field instead of the `column_index` for referencing a column.

**Technical details**

### Updated request for renaming or updating the type of columns:
```json
{
    "columns": [
        {
            "id": 1,
            "name": "id",
            "type": "INTEGER"
        },
        {
            "id": 2,
            "name": "Center",
            "type": "TEXT"
        }
    ]
}
```
### Updated request for dropping columns:
```json
{
    "columns": [
        {
            "id": 1,
            "name": "id",
            "type": "INTEGER"
        },
        {
            "id": 2
        }
    ]
}
```
## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by Anish9901** [comment]
@silentninja Can you please review this PR? I will work on refactoring failing tests once you approve the changes in the PR.

File: mathesar/utils/models.py
Comment (silentninja) [inline mathesar/utils/models.py:55]: It would be nice if column id's can be fetched from the database in a single query instead of having to fetch from the db for each column.

**Review by silentninja — CHANGES_REQUESTED** [review]
Nice work @Anish9901. I have few requests though. Also update the documentation to mention that the api has added an additional field `id`.

**Comment by silentninja** [comment]
@pavish or @seancolsen This PR brings in a breaking change and might need additional work on the frontend

**Comment by seancolsen** [comment]
I don't think we need any front end changes here. I'm reasonably certain that the front end never sends a `columns` param in a PATCH request to the tables endpoint. After some brief searching of our codebase, the closest thing I found was [here](https://github.com/centerofci/mathesar/blob/e2fedadfc1a7fe2c319e1cbf4f2cc62637779b88/mathesar_ui/src/stores/tables.ts#L142) where we update the name of a table. The `tables.ts` file seems to be the only place where we use that endpoint. And that code location seems to be the only PATCH request in that file.

**Comment by pavish** [comment]
@seancolsen @silentninja 

It is actually used in one place on the frontend: while saving the preview table. We will need frontend changes in this PR.

@seancolsen You can find the usage here: https://github.com/centerofci/mathesar/blob/e2fedadfc1a7fe2c319e1cbf4f2cc62637779b88/mathesar_ui/src/sections/import-data/importUtils.ts#L360

**Comment by pavish** [comment]
Considering that the frontend changes are supposed to be part of this PR, @Anish9901 would you be upto handling it yourself?

The core team might take a while to get to this PR.

**Comment by Anish9901** [comment]
@pavish I am not currently unfamiliar with JavaScript and Svelte so I won't be able to contribute to the frontend part. Maybe we can create a separate issue for the frontend work required for this PR.

**Comment by pavish** [comment]
@Anish9901 No worries. We'll handle the frontend changes in a day or two, but it has to be in this PR since this breaks file import at the moment. This PR can be placed on hold till then, since I also see some requested changes and failing backend tests, which might take you some time to fix.

**Comment by Anish9901** [comment]
Sure @pavish, I'll make the required changes.

**Comment by Anish9901** [comment]
@silentninja Can you please point me to the docs that need to be updated?

**Comment by Anish9901** [comment]
@kgodey @silentninja can you please change the label of this PR from waiting to review?

**Comment by Anish9901** [comment]
@silentninja If I am not wrong the import preview API spec needs change right?

**Comment by silentninja** [comment]
> If I am not wrong the import preview API spec needs change right?

Yes, right now the preview data contains `column name`, we would need to extend it to include the column `id`

**Comment by Anish9901** [comment]
@silentninja I have updated the docs can you please take a look?

File: mathesar/tests/api/test_table_api.py
Comment (silentninja) [inline mathesar/tests/api/test_table_api.py:1028]: This fixture can also be converted to accept `table_oid` and return column data  with id

**Review by silentninja — CHANGES_REQUESTED** [review]
Overall looks good, I have few minor changes that needs to be addressed.

**Review by silentninja — APPROVED** [review]

**Comment by seancolsen** [comment]
UI fixes are in #1390

**Pull request merged**
